assets/app/data/services/upgrade/openstack-services.factory.js

```diff
@@ -0,0 +1,48 @@
+(function() {
+
+    angular
+        .module('crowbarData.upgrade')
+        .factory('openStackFactory', openStackFactory);
+
+    openStackFactory.$inject = ['$q', '$http'];
+    /* @ngInject */
+    function openStackFactory($q, $http) {
+        var factory = {
+            getOpenStackServices: getOpenStackServices,
+            stopOpenstackServices: stopOpenstackServices,
+            createOpenstackBackup: createOpenstackBackup
+        };
+
+        return factory;
+
+        function getOpenStackServices() {
+
+            var requestOptions = {
+                method: 'GET',
+                url: '/api/upgrade7/openstack-services'
```

Comment:
I mean that what @santiph suggested doesn't necessarily make sense with the real backend. He wants to re-use part of the string e.g. `var x = 'a', url1 = a + '/b', url2 = a + '/c';` while the real backend can have different endpoints for those functions e.g. `var url1 = 'a/b', url2 = 'c/d';` which don't have any common part.